// tinkerforge/lib/BrickletRGBLEDButton.js
/* ***********************************************************
 * This file was automatically generated on 2019-01-29.      *
 *                                                           *
 * JavaScript Bindings Version 2.1.22                        *
 *                                                           *
 * If you have a bugfix for this file and want to commit it, *
 * please fix the bug in the generator. You can find a link  *
 * to the generators git repository on tinkerforge.com       *
 *************************************************************/

var Device = require('./Device');
var IPConnection = require('./IPConnection');

BrickletRGBLEDButton.DEVICE_IDENTIFIER = 282;
BrickletRGBLEDButton.DEVICE_DISPLAY_NAME = 'RGB LED Button Bricklet';
BrickletRGBLEDButton.CALLBACK_BUTTON_STATE_CHANGED = 4;
BrickletRGBLEDButton.FUNCTION_SET_COLOR = 1;
BrickletRGBLEDButton.FUNCTION_GET_COLOR = 2;
BrickletRGBLEDButton.FUNCTION_GET_BUTTON_STATE = 3;
BrickletRGBLEDButton.FUNCTION_SET_COLOR_CALIBRATION = 5;
BrickletRGBLEDButton.FUNCTION_GET_COLOR_CALIBRATION = 6;
BrickletRGBLEDButton.FUNCTION_GET_SPITFP_ERROR_COUNT = 234;
BrickletRGBLEDButton.FUNCTION_SET_BOOTLOADER_MODE = 235;
BrickletRGBLEDButton.FUNCTION_GET_BOOTLOADER_MODE = 236;
BrickletRGBLEDButton.FUNCTION_SET_WRITE_FIRMWARE_POINTER = 237;
BrickletRGBLEDButton.FUNCTION_WRITE_FIRMWARE = 238;
BrickletRGBLEDButton.FUNCTION_SET_STATUS_LED_CONFIG = 239;
BrickletRGBLEDButton.FUNCTION_GET_STATUS_LED_CONFIG = 240;
BrickletRGBLEDButton.FUNCTION_GET_CHIP_TEMPERATURE = 242;
BrickletRGBLEDButton.FUNCTION_RESET = 243;
BrickletRGBLEDButton.FUNCTION_WRITE_UID = 248;
BrickletRGBLEDButton.FUNCTION_READ_UID = 249;
BrickletRGBLEDButton.FUNCTION_GET_IDENTITY = 255;
BrickletRGBLEDButton.BUTTON_STATE_PRESSED = 0;
BrickletRGBLEDButton.BUTTON_STATE_RELEASED = 1;
BrickletRGBLEDButton.BOOTLOADER_MODE_BOOTLOADER = 0;
BrickletRGBLEDButton.BOOTLOADER_MODE_FIRMWARE = 1;
BrickletRGBLEDButton.BOOTLOADER_MODE_BOOTLOADER_WAIT_FOR_REBOOT = 2;
BrickletRGBLEDButton.BOOTLOADER_MODE_FIRMWARE_WAIT_FOR_REBOOT = 3;
BrickletRGBLEDButton.BOOTLOADER_MODE_FIRMWARE_WAIT_FOR_ERASE_AND_REBOOT = 4;
BrickletRGBLEDButton.BOOTLOADER_STATUS_OK = 0;
BrickletRGBLEDButton.BOOTLOADER_STATUS_INVALID_MODE = 1;
BrickletRGBLEDButton.BOOTLOADER_STATUS_NO_CHANGE = 2;
BrickletRGBLEDButton.BOOTLOADER_STATUS_ENTRY_FUNCTION_NOT_PRESENT = 3;
BrickletRGBLEDButton.BOOTLOADER_STATUS_DEVICE_IDENTIFIER_INCORRECT = 4;
BrickletRGBLEDButton.BOOTLOADER_STATUS_CRC_MISMATCH = 5;
BrickletRGBLEDButton.STATUS_LED_CONFIG_OFF = 0;
BrickletRGBLEDButton.STATUS_LED_CONFIG_ON = 1;
BrickletRGBLEDButton.STATUS_LED_CONFIG_SHOW_HEARTBEAT = 2;
BrickletRGBLEDButton.STATUS_LED_CONFIG_SHOW_STATUS = 3;

function BrickletRGBLEDButton(uid, ipcon) {
	//Push button with built-in RGB LED

	/*
	Creates an object with the unique device ID *uid* and adds it to
	the IP Connection *ipcon*.
	*/
	Device.call(this, this, uid, ipcon);
	BrickletRGBLEDButton.prototype = Object.create(Device);
	this.APIVersion = [2, 0, 0];
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_SET_COLOR] = Device.RESPONSE_EXPECTED_FALSE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_GET_COLOR] = Device.RESPONSE_EXPECTED_ALWAYS_TRUE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_GET_BUTTON_STATE] = Device.RESPONSE_EXPECTED_ALWAYS_TRUE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_SET_COLOR_CALIBRATION] = Device.RESPONSE_EXPECTED_FALSE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_GET_COLOR_CALIBRATION] = Device.RESPONSE_EXPECTED_ALWAYS_TRUE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_GET_SPITFP_ERROR_COUNT] = Device.RESPONSE_EXPECTED_ALWAYS_TRUE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_SET_BOOTLOADER_MODE] = Device.RESPONSE_EXPECTED_ALWAYS_TRUE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_GET_BOOTLOADER_MODE] = Device.RESPONSE_EXPECTED_ALWAYS_TRUE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_SET_WRITE_FIRMWARE_POINTER] = Device.RESPONSE_EXPECTED_FALSE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_WRITE_FIRMWARE] = Device.RESPONSE_EXPECTED_ALWAYS_TRUE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_SET_STATUS_LED_CONFIG] = Device.RESPONSE_EXPECTED_FALSE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_GET_STATUS_LED_CONFIG] = Device.RESPONSE_EXPECTED_ALWAYS_TRUE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_GET_CHIP_TEMPERATURE] = Device.RESPONSE_EXPECTED_ALWAYS_TRUE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_RESET] = Device.RESPONSE_EXPECTED_FALSE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_WRITE_UID] = Device.RESPONSE_EXPECTED_FALSE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_READ_UID] = Device.RESPONSE_EXPECTED_ALWAYS_TRUE;
	this.responseExpected[BrickletRGBLEDButton.FUNCTION_GET_IDENTITY] = Device.RESPONSE_EXPECTED_ALWAYS_TRUE;
	this.callbackFormats[BrickletRGBLEDButton.CALLBACK_BUTTON_STATE_CHANGED] = 'B';



	this.setColor = function(red, green, blue, returnCallback, errorCallback) {
		/*
		Sets the color of the LED.
		
		By default the LED is off (0, 0, 0).
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_SET_COLOR, [red, green, blue], 'B B B', '', returnCallback, errorCallback, false);
	};
	this.getColor = function(returnCallback, errorCallback) {
		/*
		Returns the LED color as set by :func:`Set Color`.
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_GET_COLOR, [], '', 'B B B', returnCallback, errorCallback, false);
	};
	this.getButtonState = function(returnCallback, errorCallback) {
		/*
		Returns the current state of the button (either pressed or released).
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_GET_BUTTON_STATE, [], '', 'B', returnCallback, errorCallback, false);
	};
	this.setColorCalibration = function(red, green, blue, returnCallback, errorCallback) {
		/*
		Sets a color calibration. Some colors appear brighter then others,
		so a calibration may be necessary for nice uniform colors.
		
		The values range from 0% to 100%.
		
		The calibration is saved in flash. You don't need to call this
		function on every startup.
		
		Default value is (100, 100, 55).
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_SET_COLOR_CALIBRATION, [red, green, blue], 'B B B', '', returnCallback, errorCallback, false);
	};
	this.getColorCalibration = function(returnCallback, errorCallback) {
		/*
		Returns the color calibration as set by :func:`Set Color Calibration`.
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_GET_COLOR_CALIBRATION, [], '', 'B B B', returnCallback, errorCallback, false);
	};
	this.getSPITFPErrorCount = function(returnCallback, errorCallback) {
		/*
		Returns the error count for the communication between Brick and Bricklet.
		
		The errors are divided into
		
		* ACK checksum errors,
		* message checksum errors,
		* framing errors and
		* overflow errors.
		
		The errors counts are for errors that occur on the Bricklet side. All
		Bricks have a similar function that returns the errors on the Brick side.
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_GET_SPITFP_ERROR_COUNT, [], '', 'I I I I', returnCallback, errorCallback, false);
	};
	this.setBootloaderMode = function(mode, returnCallback, errorCallback) {
		/*
		Sets the bootloader mode and returns the status after the requested
		mode change was instigated.
		
		You can change from bootloader mode to firmware mode and vice versa. A change
		from bootloader mode to firmware mode will only take place if the entry function,
		device identifier and CRC are present and correct.
		
		This function is used by Brick Viewer during flashing. It should not be
		necessary to call it in a normal user program.
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_SET_BOOTLOADER_MODE, [mode], 'B', 'B', returnCallback, errorCallback, false);
	};
	this.getBootloaderMode = function(returnCallback, errorCallback) {
		/*
		Returns the current bootloader mode, see :func:`Set Bootloader Mode`.
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_GET_BOOTLOADER_MODE, [], '', 'B', returnCallback, errorCallback, false);
	};
	this.setWriteFirmwarePointer = function(pointer, returnCallback, errorCallback) {
		/*
		Sets the firmware pointer for :func:`Write Firmware`. The pointer has
		to be increased by chunks of size 64. The data is written to flash
		every 4 chunks (which equals to one page of size 256).
		
		This function is used by Brick Viewer during flashing. It should not be
		necessary to call it in a normal user program.
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_SET_WRITE_FIRMWARE_POINTER, [pointer], 'I', '', returnCallback, errorCallback, false);
	};
	this.writeFirmware = function(data, returnCallback, errorCallback) {
		/*
		Writes 64 Bytes of firmware at the position as written by
		:func:`Set Write Firmware Pointer` before. The firmware is written
		to flash every 4 chunks.
		
		You can only write firmware in bootloader mode.
		
		This function is used by Brick Viewer during flashing. It should not be
		necessary to call it in a normal user program.
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_WRITE_FIRMWARE, [data], 'B64', 'B', returnCallback, errorCallback, false);
	};
	this.setStatusLEDConfig = function(config, returnCallback, errorCallback) {
		/*
		Sets the status LED configuration. By default the LED shows
		communication traffic between Brick and Bricklet, it flickers once
		for every 10 received data packets.
		
		You can also turn the LED permanently on/off or show a heartbeat.
		
		If the Bricklet is in bootloader mode, the LED is will show heartbeat by default.
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_SET_STATUS_LED_CONFIG, [config], 'B', '', returnCallback, errorCallback, false);
	};
	this.getStatusLEDConfig = function(returnCallback, errorCallback) {
		/*
		Returns the configuration as set by :func:`Set Status LED Config`
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_GET_STATUS_LED_CONFIG, [], '', 'B', returnCallback, errorCallback, false);
	};
	this.getChipTemperature = function(returnCallback, errorCallback) {
		/*
		Returns the temperature in °C as measured inside the microcontroller. The
		value returned is not the ambient temperature!
		
		The temperature is only proportional to the real temperature and it has bad
		accuracy. Practically it is only useful as an indicator for
		temperature changes.
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_GET_CHIP_TEMPERATURE, [], '', 'h', returnCallback, errorCallback, false);
	};
	this.reset = function(returnCallback, errorCallback) {
		/*
		Calling this function will reset the Bricklet. All configurations
		will be lost.
		
		After a reset you have to create new device objects,
		calling functions on the existing ones will result in
		undefined behavior!
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_RESET, [], '', '', returnCallback, errorCallback, false);
	};
	this.writeUID = function(uid, returnCallback, errorCallback) {
		/*
		Writes a new UID into flash. If you want to set a new UID
		you have to decode the Base58 encoded UID string into an
		integer first.
		
		We recommend that you use Brick Viewer to change the UID.
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_WRITE_UID, [uid], 'I', '', returnCallback, errorCallback, false);
	};
	this.readUID = function(returnCallback, errorCallback) {
		/*
		Returns the current UID as an integer. Encode as
		Base58 to get the usual string version.
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_READ_UID, [], '', 'I', returnCallback, errorCallback, false);
	};
	this.getIdentity = function(returnCallback, errorCallback) {
		/*
		Returns the UID, the UID where the Bricklet is connected to,
		the position, the hardware and firmware version as well as the
		device identifier.
		
		The position can be 'a', 'b', 'c' or 'd'.
		
		The device identifier numbers can be found :ref:`here <device_identifier>`.
		|device_identifier_constant|
		*/
		this.ipcon.sendRequest(this, BrickletRGBLEDButton.FUNCTION_GET_IDENTITY, [], '', 's8 s8 c B3 B3 H', returnCallback, errorCallback, false);
	};
}

module.exports = BrickletRGBLEDButton;
