// tinkerforge/Tinkerforge.js
function Tinkerforge() {
	this.IPConnection = require('./lib/IPConnection');
	this.BrickDC = require('./lib/BrickDC');
	this.BrickIMU = require('./lib/BrickIMU');
	this.BrickIMUV2 = require('./lib/BrickIMUV2');
	this.BrickMaster = require('./lib/BrickMaster');
	this.BrickRED = require('./lib/BrickRED');
	this.BrickServo = require('./lib/BrickServo');
	this.BrickSilentStepper = require('./lib/BrickSilentStepper');
	this.BrickStepper = require('./lib/BrickStepper');
	this.BrickletAccelerometer = require('./lib/BrickletAccelerometer');
	this.BrickletAccelerometerV2 = require('./lib/BrickletAccelerometerV2');
	this.BrickletAirQuality = require('./lib/BrickletAirQuality');
	this.BrickletAmbientLight = require('./lib/BrickletAmbientLight');
	this.BrickletAmbientLightV2 = require('./lib/BrickletAmbientLightV2');
	this.BrickletAmbientLightV3 = require('./lib/BrickletAmbientLightV3');
	this.BrickletAnalogIn = require('./lib/BrickletAnalogIn');
	this.BrickletAnalogInV2 = require('./lib/BrickletAnalogInV2');
	this.BrickletAnalogInV3 = require('./lib/BrickletAnalogInV3');
	this.BrickletAnalogOut = require('./lib/BrickletAnalogOut');
	this.BrickletAnalogOutV2 = require('./lib/BrickletAnalogOutV2');
	this.BrickletAnalogOutV3 = require('./lib/BrickletAnalogOutV3');
	this.BrickletBarometer = require('./lib/BrickletBarometer');
	this.BrickletBarometerV2 = require('./lib/BrickletBarometerV2');
	this.BrickletCAN = require('./lib/BrickletCAN');
	this.BrickletCANV2 = require('./lib/BrickletCANV2');
	this.BrickletCO2 = require('./lib/BrickletCO2');
	this.BrickletColor = require('./lib/BrickletColor');
	this.BrickletCurrent12 = require('./lib/BrickletCurrent12');
	this.BrickletCurrent25 = require('./lib/BrickletCurrent25');
	this.BrickletDistanceIR = require('./lib/BrickletDistanceIR');
	this.BrickletDistanceIRV2 = require('./lib/BrickletDistanceIRV2');
	this.BrickletDistanceUS = require('./lib/BrickletDistanceUS');
	this.BrickletDMX = require('./lib/BrickletDMX');
	this.BrickletDualButton = require('./lib/BrickletDualButton');
	this.BrickletDualButtonV2 = require('./lib/BrickletDualButtonV2');
	this.BrickletDualRelay = require('./lib/BrickletDualRelay');
	this.BrickletDustDetector = require('./lib/BrickletDustDetector');
	this.BrickletGPS = require('./lib/BrickletGPS');
	this.BrickletGPSV2 = require('./lib/BrickletGPSV2');
	this.BrickletHallEffect = require('./lib/BrickletHallEffect');
	this.BrickletHumidity = require('./lib/BrickletHumidity');
	this.BrickletHumidityV2 = require('./lib/BrickletHumidityV2');
	this.BrickletIndustrialAnalogOut = require('./lib/BrickletIndustrialAnalogOut');
	this.BrickletIndustrialAnalogOutV2 = require('./lib/BrickletIndustrialAnalogOutV2');
	this.BrickletIndustrialCounter = require('./lib/BrickletIndustrialCounter');
	this.BrickletIndustrialDigitalIn4 = require('./lib/BrickletIndustrialDigitalIn4');
	this.BrickletIndustrialDigitalIn4V2 = require('./lib/BrickletIndustrialDigitalIn4V2');
	this.BrickletIndustrialDigitalOut4 = require('./lib/BrickletIndustrialDigitalOut4');
	this.BrickletIndustrialDigitalOut4V2 = require('./lib/BrickletIndustrialDigitalOut4V2');
	this.BrickletIndustrialDual020mA = require('./lib/BrickletIndustrialDual020mA');
	this.BrickletIndustrialDual020mAV2 = require('./lib/BrickletIndustrialDual020mAV2');
	this.BrickletIndustrialDualAnalogIn = require('./lib/BrickletIndustrialDualAnalogIn');
	this.BrickletIndustrialDualAnalogInV2 = require('./lib/BrickletIndustrialDualAnalogInV2');
	this.BrickletIndustrialDualRelay = require('./lib/BrickletIndustrialDualRelay');
	this.BrickletIndustrialQuadRelay = require('./lib/BrickletIndustrialQuadRelay');
	this.BrickletIndustrialQuadRelayV2 = require('./lib/BrickletIndustrialQuadRelayV2');
	this.BrickletIO16 = require('./lib/BrickletIO16');
	this.BrickletIO16V2 = require('./lib/BrickletIO16V2');
	this.BrickletIO4 = require('./lib/BrickletIO4');
	this.BrickletIO4V2 = require('./lib/BrickletIO4V2');
	this.BrickletIsolator = require('./lib/BrickletIsolator');
	this.BrickletJoystick = require('./lib/BrickletJoystick');
	this.BrickletLaserRangeFinder = require('./lib/BrickletLaserRangeFinder');
	this.BrickletLCD128x64 = require('./lib/BrickletLCD128x64');
	this.BrickletLCD16x2 = require('./lib/BrickletLCD16x2');
	this.BrickletLCD20x4 = require('./lib/BrickletLCD20x4');
	this.BrickletLEDStrip = require('./lib/BrickletLEDStrip');
	this.BrickletLEDStripV2 = require('./lib/BrickletLEDStripV2');
	this.BrickletLine = require('./lib/BrickletLine');
	this.BrickletLinearPoti = require('./lib/BrickletLinearPoti');
	this.BrickletLoadCell = require('./lib/BrickletLoadCell');
	this.BrickletLoadCellV2 = require('./lib/BrickletLoadCellV2');
	this.BrickletMoisture = require('./lib/BrickletMoisture');
	this.BrickletMotionDetector = require('./lib/BrickletMotionDetector');
	this.BrickletMotionDetectorV2 = require('./lib/BrickletMotionDetectorV2');
	this.BrickletMotorizedLinearPoti = require('./lib/BrickletMotorizedLinearPoti');
	this.BrickletMultiTouch = require('./lib/BrickletMultiTouch');
	this.BrickletNFC = require('./lib/BrickletNFC');
	this.BrickletNFCRFID = require('./lib/BrickletNFCRFID');
	this.BrickletOLED128x64 = require('./lib/BrickletOLED128x64');
	this.BrickletOLED128x64V2 = require('./lib/BrickletOLED128x64V2');
	this.BrickletOLED64x48 = require('./lib/BrickletOLED64x48');
	this.BrickletOneWire = require('./lib/BrickletOneWire');
	this.BrickletOutdoorWeather = require('./lib/BrickletOutdoorWeather');
	this.BrickletParticulateMatter = require('./lib/BrickletParticulateMatter');
	this.BrickletPiezoBuzzer = require('./lib/BrickletPiezoBuzzer');
	this.BrickletPiezoSpeaker = require('./lib/BrickletPiezoSpeaker');
	this.BrickletPTC = require('./lib/BrickletPTC');
	this.BrickletPTCV2 = require('./lib/BrickletPTCV2');
	this.BrickletRealTimeClock = require('./lib/BrickletRealTimeClock');
	this.BrickletRealTimeClockV2 = require('./lib/BrickletRealTimeClockV2');
	this.BrickletRemoteSwitch = require('./lib/BrickletRemoteSwitch');
	this.BrickletRemoteSwitchV2 = require('./lib/BrickletRemoteSwitchV2');
	this.BrickletRGBLEDButton = require('./lib/BrickletRGBLEDButton');
	this.BrickletRGBLED = require('./lib/BrickletRGBLED');
	this.BrickletRGBLEDMatrix = require('./lib/BrickletRGBLEDMatrix');
	this.BrickletRotaryEncoder = require('./lib/BrickletRotaryEncoder');
	this.BrickletRotaryEncoderV2 = require('./lib/BrickletRotaryEncoderV2');
	this.BrickletRotaryPoti = require('./lib/BrickletRotaryPoti');
	this.BrickletRS232 = require('./lib/BrickletRS232');
	this.BrickletRS232V2 = require('./lib/BrickletRS232V2');
	this.BrickletRS485 = require('./lib/BrickletRS485');
	this.BrickletSegmentDisplay4x7 = require('./lib/BrickletSegmentDisplay4x7');
	this.BrickletSolidStateRelay = require('./lib/BrickletSolidStateRelay');
	this.BrickletSolidStateRelayV2 = require('./lib/BrickletSolidStateRelayV2');
	this.BrickletSoundIntensity = require('./lib/BrickletSoundIntensity');
	this.BrickletSoundPressureLevel = require('./lib/BrickletSoundPressureLevel');
	this.BrickletTemperature = require('./lib/BrickletTemperature');
	this.BrickletTemperatureIR = require('./lib/BrickletTemperatureIR');
	this.BrickletTemperatureIRV2 = require('./lib/BrickletTemperatureIRV2');
	this.BrickletTemperatureV2 = require('./lib/BrickletTemperatureV2');
	this.BrickletThermalImaging = require('./lib/BrickletThermalImaging');
	this.BrickletThermocouple = require('./lib/BrickletThermocouple');
	this.BrickletThermocoupleV2 = require('./lib/BrickletThermocoupleV2');
	this.BrickletTilt = require('./lib/BrickletTilt');
	this.BrickletUVLight = require('./lib/BrickletUVLight');
	this.BrickletUVLightV2 = require('./lib/BrickletUVLightV2');
	this.BrickletVoltage = require('./lib/BrickletVoltage');
	this.BrickletVoltageCurrent = require('./lib/BrickletVoltageCurrent');
	this.BrickletVoltageCurrentV2 = require('./lib/BrickletVoltageCurrentV2');
}

module.exports = new Tinkerforge();
